src/python/pants/java/nailgun_io.py

```diff
@@ -16,21 +16,46 @@
 from pants.java.nailgun_protocol import ChunkType, NailgunProtocol
 
 
-@contextmanager
-def _pipe(isatty):
-  r_fd, w_fd = os.openpty() if isatty else os.pipe()
-  yield (r_fd, w_fd)
+class Pipe(object):
 
+  def __init__(self, read_fd, write_fd):
+    self.read_fd = read_fd
+    self.write_fd = write_fd
 
-@contextmanager
-def _self_closing_pipe(isatty):
-  with _pipe(isatty) as (r_fd, w_fd):
+  def is_writable(self):
+    """
+    If the write end of a pipe closes, the read end might still be open, to allow
+    readers to finish reading before closing it.
+    However, there are cases where we still want to know if the write end is closed.
+
+    :return: True if the write end of the pipe is open.
+    """
+    try:
+      os.fstat(self.write_fd)
+    except OSError:
+      return False
+    return True
+
+  def close(self):
+    """Close the reading end of the pipe, which should close the writing end too."""
+    os.close(self.read_fd)
+
+  @staticmethod
+  def open(isatty):
```

Comment:
Shadowing a builtin is not recommend, as in certain scopes it means we can't use the builtin `open()` anymore.

Maybe `open_pipe`? I know the namespace is annoying to say `Pipe.open_pipe()`, but worth it to avoid shadowing imo. Or, you could call it `Pipe.create_and_open()`? `Pipe.create()`?

Reply:
Just making this a `@classmethod` instead of `@staticmethod` would do the job here, right?

Reply:
> Just making this a @classmethod instead of @staticmethod would do the job here, right?

Wouldn't have an impact.

Although while writing out the response to this, I realized that actually we will never have issues with shadowing!

The reason this use is okay is that `def open()` is attached to the class, rather than being a "free" function defined at the file's top level. The only way to reference this function is by first referencing the class, e.g. `Pipe.open()`. This is true even when code is located in the `Pipe` class, e.g. when defining other methods. So, `.open()` now becomes a property defined on `Pipe`, rather than shadowing the unbound builtin `open()`.

So this if fine to keep, although perhaps confusing.